clash-cores/src/Clash/Cores/Xilinx/VIO/Internal/BlackBoxes.hs

```diff
@@ -98,42 +102,67 @@ vioProbeBBTF :: Backend s => BlackBoxContext -> State s Doc
 vioProbeBBTF bbCtx
   | (   _knownDomainDom
       : _vioConstraint
+      : (DSL.getVec -> Just inputNameExprs)
+      : (DSL.getVec -> Just outputNameExprs)
       : _initialOutputProbeValues
       : clk
       : inputProbes
       ) <- map fst $ DSL.tInputs bbCtx
+  , Just inputNames <- sequence (map (fmap T.pack . DSL.getStr) inputNameExprs)
+  , Just outputNames <- sequence (map (fmap T.pack . DSL.getStr) outputNameExprs)
   , [vioProbeName] <- bbQsysIncName bbCtx
   , Right (inTys, outTys) <- probesFromTypes bbCtx
   , [tResult] <- map DSL.ety (DSL.tResults bbCtx)
   = do
+      if length inTys /= length inputNames then
+        error [__i|
+          Number of input names did not match number of input probes. Expected
+          #{length inTys} input name(s), got #{length inputNames}. Got input name(s):
+
+            #{ppShow inputNames}
+        |]
+      else if length outTys /= length outputNames then
+        error [__i|
+          Number of output names did not match number of output probes. Expected
+          #{length outTys} output name(s), got #{length outputNames}. Got output name(s):
+
+            #{ppShow outputNames}
+        |]
+      else
+        pure ()
+
       vioProbeInstName <- Id.makeBasic "vio_inst"
 
       let
         inPs = filter ((> (0 :: Int)) . DSL.tySize . DSL.ety) inputProbes
         inNames = map (T.pack . ("probe_in" <>) . show) [0 :: Int, 1..]
         outNames = map (T.pack . ("probe_out" <>) . show) [0 :: Int, 1..]
-        inBVNames = map (`T.append` "_bv") inNames
-        outBVNames = map (`T.append` "_bv") outNames
+        attrs = [StringAttr' "KEEP" "true"]
```

Comment:
Could you add a comment clarifying why wee need `KEEP`?

Reply:
Also, in the commit message, please use American English. (I.e., `optimization` instead of `optimisation`.)